hoot-rnd/src/main/cpp/hoot/rnd/conflate/multiary/MultiaryPoiMergeCache.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2017 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+#include "MultiaryPoiMergeCache.h"
+
+// hoot
+#include <hoot/core/conflate/MatchCreator.h>
+#include <hoot/core/conflate/Merger.h>
+#include <hoot/core/conflate/MergerCreator.h>
+#include <hoot/core/ops/CopySubsetOp.h>
+
+namespace hoot
+{
+
+/**
+ * Order first by input number, then by element ID. This should always give a consistent ordering
+ * even if the elements come from the same input.
+ */
+struct ElementOrder
+{
+public:
+  bool operator()(const ConstElementPtr& e1, const ConstElementPtr& e2)
+  {
+    if (e1->getStatus().getInput() < e2->getStatus().getInput())
+    {
+      return true;
+    }
+    else
+    {
+      return e1->getElementId() < e2->getElementId();
+    }
+  }
+};
+
+/**
+ * Delete the MatchSet structure in a safe way.
+ */
+struct DeleteMatches
+{
+  DeleteMatches(MatchSet& ms) : _ms(ms) {}
+
+  ~DeleteMatches()
+  {
+    foreach (const Match* m, _ms)
+    {
+      delete m;
+    }
+  }
+
+  MatchSet _ms;
+};
+
+/**
+ * Delete the Merger vector in a safe way.
+ */
```

Comment:
Good catch!

It should definitely be passed by reference, but otherwise the destructor should do the trick. This is similar to the way auto_ptr deletes things. The order of destruction should guarantee that the DeleteMergers destructor is called before the vector destructor.

I agree, shared_ptr would be better. These vectors of pointers give me the willies. (Puts on old man cap and uses Abraham Simpson's voice) Back in the good ol' days, I didn't use shared pointers here for performance reasons. We have since disabled thread support for shared_ptr and it isn't an issue. However, to replace this with a vector of shared_ptr would require some significant.